staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1467 @@
+		// to remote. After doing this, we check whether the local attachment
+		// we've removed was the last one for its vni. If that's the case, the
+		// vni is no longer relevant to the connection agent, hence we unset and
+		// return the vnState so that the caller can perform the necessary clean
+		// up. If the vnState is missing, we set the return flag
+		// noVnStateFoundForRemoteAtt to true so that the caller knows that the
+		// remote attachment it is processing must not get an interface (this is
+		// needed because a reference to such attachment is not necessarily
+		// already in the list of remote attachments in the vn, hence it's not
+		// granted that such a reference has been enqueued to delete the
+		// attachment).
+		if vnState != nil {
+			vnState.remoteAtts[attName] = struct{}{}
+			delete(vnState.localAtts, attName)
+			if len(vnState.localAtts) == 0 {
+				delete(ca.vniToVnState, vni)
+				vnStateRet = vnState
+			}
+		} else {
+			noVnStateFoundForRemoteAtt = true
+		}
+	}
+
+	return
+}
+
+func (ca *ConnectionAgent) updateVNStateAfterAttDeparture(attName string, vni uint32) {
+	vnState := ca.removeAttFromVNState(attName, vni)
+	if vnState == nil {
+		return
+	}
+	// If we're here attName was the last local attachment in the virtual network
+	// with id vni. Hence we stop the remote attachments informer and enqueue
+	// references to remote attachments in that virtual network, so that their
+	// interfaces can be deleted.
+	ca.clearVNResources(vnState, attName, vni)
+}
+
+// createOrUpdateIfc returns the MAC address, Linux interface name,
+// post-create exec report, and possibly an error
+func (ca *ConnectionAgent) createOrUpdateIfc(attGuestIP, attHostIP gonet.IP, attVNI uint32, att *netv1a1.NetworkAttachment) (attMACStr, ifcName string, statusErrs SliceOfString, pcer *netv1a1.ExecReport, err error) {
+	attNSN := parse.AttNSN(att)
+	attMAC := generateMACAddr(attVNI, attGuestIP)
+	attMACStr = attMAC.String()
+	ifcName = generateIfcName(attMAC)
+	oldLocalState, attHasLocalIfc := ca.getLocalAttState(attNSN)
+	oldRemoteIfc, attHasRemoteIfc := ca.getRemoteIfc(attNSN)
+	newIfcNeedsToBeCreated := (!attHasLocalIfc && !attHasRemoteIfc) ||
+		(attHasLocalIfc && ifcNeedsUpdate(ca.hostIP, attHostIP, oldLocalState.GuestMAC, attMAC)) ||
+		(attHasRemoteIfc && ifcNeedsUpdate(oldRemoteIfc.HostIP, attHostIP, oldRemoteIfc.GuestMAC, attMAC))
+
+	if newIfcNeedsToBeCreated {
+		if attHasLocalIfc {
+			if err := ca.DeleteLocalIfc(oldLocalState.LocalNetIfc); err != nil {
+				return "", "", nil, nil, fmt.Errorf("update of network interface of attachment %s failed, old local interface %#+v could not be deleted: %s",
+					attNSN,
+					oldLocalState.LocalNetIfc,
+					err.Error())
+			}
+			ca.LaunchCommand(attNSN, &oldLocalState.LocalNetIfc, oldLocalState.PostDeleteExec, "postDelete", true, false)
+			ca.unsetLocalAttState(attNSN)
+		}
+		if attHasRemoteIfc {
+			if err := ca.DeleteRemoteIfc(oldRemoteIfc); err != nil {
```

Comment:
> I like the idea of not pushing the difficulty into the fabric. Local and remote are significantly different in their network implementations, so it is natural for the fabric to have different APIs for them.

Complexity of the implementation of the network fabric code is not much: just separate local and remote interfaces using `hostIP` and separate processing from there. Most of the current network fabric implementation code would be unchanged. For instance, there would be an exported `CreateNetworkInterface(ifc NetworkInterface)` function. Internally it would have an if which depending on the `hostIP` of `ifc` would invoke `createLocalNetworkInterface` or `createRemoteNetworkInterface`, which would be almost the exact same functions we have now, only made private.

But I agree that the semantic difference between local and remote network interfaces is big enough to make it desirable to have separate APIs.  OTOH I really don't like how the distinction turns out in the connection agent code. Maybe with the solution you outlined above and with some extra functions in the connection agent to hide from the main logic the difference between API calls for local and remote network interfaces we can make it simpler and more readable. I'm pausing now, will think about this more later.

Reply:
Ok, I think I have a rough idea on how to do this. I'll start coding stuff for which we've reached consensus.